### DISCUSSION TITLE
[Propousal] Readable using statments

### DISCUSSION THREAD

**Discussion by redradist** [dotnet/csharplang#3048] [Language Ideas]
When we want to use some library we have to add using statement:
```csharp
using System;
using System.Threading;

public class ThreadExample {
    public static void ThreadProc() {
        for (int i = 0; i < 10; i++) {
            Console.WriteLine("ThreadProc: {0}", i);
            Thread.Sleep(0);
        }
    }

    public static void Main() {
        Console.WriteLine("Main thread: Start a second thread.");
        Thread t = new Thread(new ThreadStart(ThreadProc));

        t.Start();

        for (int i = 0; i < 4; i++) {
            Console.WriteLine("Main thread: Do some work.");
            Thread.Sleep(0);
        }

        Console.WriteLine("Main thread: Call Join(), to wait until ThreadProc ends.");
        t.Join();
        Console.WriteLine("Main thread: ThreadProc.Join has returned.  Press Enter to end program.");
        Console.ReadLine();
    }
}
```

but ... it make hard reading and learning from the code ... (

Consider to enhancement using by default the following syntax:
```csharp
using System;
using System.Threading::{Thread, ThreadStart};

public class ThreadExample {
    public static void ThreadProc() {
        for (int i = 0; i < 10; i++) {
            Console.WriteLine("ThreadProc: {0}", i);
            Thread.Sleep(0);
        }
    }

    public static void Main() {
        Console.WriteLine("Main thread: Start a second thread.");
        Thread t = new Thread(new ThreadStart(ThreadProc));

        t.Start();

        for (int i = 0; i < 4; i++) {
            Console.WriteLine("Main thread: Do some work.");
            Thread.Sleep(0);
        }

        Console.WriteLine("Main thread: Call Join(), to wait until ThreadProc ends.");
        t.Join();
        Console.WriteLine("Main thread: ThreadProc.Join has returned.  Press Enter to end program.");
        Console.ReadLine();
    }
}
```

Share your opinion ...
Also be aware that this feature is taken from such languages as Java, Python, Rust, Javascript ...
And it could mean that this feature has been proved as useful ;)

**Comment by QuantumDeveloper** [comment]
For me as for JS hater this syntax look odd. 
You can always go inside C# dll and see all types inside. I dont want to explicitly set all types each time I add some using. 

PS. Some languages does not have namespaces at all (Objective-C, Swift) so why not to just remove namespaces from C#? (sarcasm)

**Comment by khm1600** [comment]
Similar: #452 #2078

**Comment by huoyaoyuan** [comment]
> it make hard learning from the code

There's nothing to learn here. **Never** memorize which type lives in which namespace. The tooling today also discourage this: when you hover on a type name, the tooltip will show its full name. When you typing an unimported type, the importer will provide the namespace for you.

> And it could mean that this feature has been proved as useful ;)

It's strongly depending on the organization of standard library. .NET's standard library is organized in multi-level hierarchy. Using a namespace only import a group of types.


Given the sample case, writing that `Thread` lives in `System.Threading` gives 0 information. It's just institutive.

**Comment by redradist** [comment]
> > it make hard learning from the code
> 
> There's nothing to learn here. **Never** memorize which type lives in which namespace. The tooling today also discourage this: when you hover on a type name, the tooltip will show its full name. When you typing an unimported type, the importer will provide the namespace for you.
> 
> > And it could mean that this feature has been proved as useful ;)
> 
> It's strongly depending on the organization of standard library. .NET's standard library is organized in multi-level hierarchy. Using a namespace only import a group of types.
> 
> Given the sample case, writing that `Thread` lives in `System.Threading` gives 0 information. It's just institutive.

What if you are viewing code on GitHub ? What kind if tool will help you ? Code should be readable by default without any external tool ... Otherwise we can put in tool all features of language !!!!!

**Comment by huoyaoyuan** [comment]
> What if you are viewing code on GitHub ? What kind if tool will help you ?

When reading code on GitHub, I never look at the `using`s. The type name `Thread` is enough.

**Comment by redradist** [comment]
@huoyaoyuan  
> > What if you are viewing code on GitHub ? What kind if tool will help you ?
> 
> When reading code on GitHub, I never look at the `using`s. The type name `Thread` is enough.

It is not about obvious things like Thread in System.Threading, but about readability !!
Imagine more overloaded namespace and multiple types with similar names that leave in different namespaces !!

Looking at code in C# you never now where is type located !!

**Comment by theunrepentantgeek** [comment]
> Looking at code in C# you never now where is type located !!

I’ve worked on multiple large .Net projects (large meaning >1M loc) and I can literally count on one hand the number of times this has been an issue.

It’s also worth noting that the C# **using** statement ***is not*** about importing types, it’s about convenience - it’s always possible to use the fully qualified type name, once the project has an assembly reference. This makes it very different from languages like Javascript and Typescript.

**Comment by redradist** [comment]
@theunrepentantgeek 
> It’s also worth noting that the C# **using** statement _**is not**_ about importing types, it’s about convenience - it’s always possible to use the fully qualified type name, once the project has an assembly reference.

But if it would be possible to show which types are used in file than it will improve readability !!

@theunrepentantgeek 
> This makes it very different from languages like Javascript and Typescript.

If some feature is different it does not mean that it is better ;)

**Comment by HaloFour** [comment]
> Also be aware that this feature is taken from such languages as Java

The language team was well aware of how `import` statements worked in Java and they decided against implementing namespaces like packages.

C# does not work the way the languages you referenced work.  You're not importing specific types from a module or package.  You're working with namespaces, and bringing them into scope.  This is exactly how `using namespace` works in C++.  I don't see why C# needs to reverse course on this, especially since it has not been a point of contention over the last 20 years and this proposal would only serve to make the language more verbose.  I severely doubt the vast majority of developers would bother using it, so it would not bring any perceived "readability" benefits.

**Comment by redradist** [comment]
@HaloFour 
> This is exactly how `using namespace` works in C++

And in world of C++ it is considered to be antipattern using whole namespace

@HaloFour 
> I severely doubt the vast majority of developers would bother using it, so it would not bring any perceived "readability" benefits.

If Visual Studio will add by default a new using statement than during the time it will be possible to change situation 😉

**Comment by theunrepentantgeek** [comment]
>If some feature is different it does not mean that it is better ;)

True enough. Except that you're arguing that other languages do this thing you want, so C# should do it too. But, those languages don't do the thing you think they do, so the analogy is weak.

> But if it would be possible to show which types are used in file than it will improve readability !!

Readability is a _subjective_ measure - a matter of opinion & familiarity.

For example, when I'm working on Angular projects, all the extra stuff at the top of the file just looks to me like a mess - it's all cluttered and difficult to read. Having to explicitly list each separate type that I want to use feels like useless boilerplate.

This doesn't mean that it's wrong for TypeScript to explicitly include the types; just that it's different - and that one person's _improved readability_ is another's _cluttered mess_.

Instead of arguing that it improves readability, you need to demonstrate how this solves real problems that are commonly encountered by other developers.

**Comment by HaloFour** [comment]
@redradist 

> And in world of C++ it is considered to be antipattern using whole namespace

Only for `std` and only because it's massive and flat.  .NET does not suffer from that problem.  It's also an anti-pattern to import namespaces in header files since that impacts source files that include said header, which is also not a problem in C# since `using` directives are always scoped to the current source file and never impact other files.

https://stackoverflow.com/a/26722134/2900606

> If Visual Studio will add by default a new using statement than during the time it will be possible to change situation

That will certainly not happen.  At best you could hope for would be an opt-in formatting option, which most developers wouldn't enable (or want to enable).

**Comment by spydacarnage** [comment]
This sounds like something that could be written as an extension, like
CodeLens, but I certainly don't think it should be part of the language.

**Comment by Summpot** [comment]
I think this approach is feasible, sometimes you can’t get the context of the code or assembly, for example when browsing the code on github, but the syntax is a bit ugly, there may be ways to improve.May be`using { ... } in ...` is ok? Need more discussion.

**Comment by redradist** [comment]
@Summpot 
> I think this approach is feasible, sometimes you can’t get the context of the code or assembly, for example when browsing the code on github, but the syntax is a bit ugly, there may be ways to improve.May be`using { ... } in ...` is ok? Need more discussion.

From my point of view it is bad, 'cause it is not obvious the hierarchy of class in namespaces

**Comment by CyrusNajmabadi** [comment]
It also doesn't work in left-to-right intellisense.

**Comment by Summpot** [comment]
@redradist 
I am sorry for the bad syntax I proposed, but I think this feature is worth discussing, there are many things we have not considered, such as extension methods, and in the case of ILLinker. It may be beneficial to introduce another keyword `as` , on the one hand, it can replace the ugly `=` in the using block (name conflicts are really common), on the other hand, it no longer needs the support of IntelliSense, because its Usage will be similar to `var` . But at the moment I can't think of any good ideas in specific syntax.
